zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+    // to have been mined in a transaction that we currently see as unmined.
+    // This is conservative in terms of avoiding violation of the gap
+    // invariant: it can only cause us to get to the end of the gap sooner.
+    let last_mined_index: i32 = match conn
+        .query_row(
+            "SELECT address_index FROM ephemeral_addresses
+             JOIN transactions t ON t.id_tx = mined_in_tx
+             WHERE account_id = :account_id AND t.mined_height IS NOT NULL
+             ORDER BY address_index DESC
+             LIMIT 1",
+            named_params![":account_id": account_id.0],
+            |row| row.get::<_, i32>(0),
+        )
+        .optional()?
+    {
+        Some(i) if i < 0 => Err(SqliteClientError::CorruptedData(
+            "negative index".to_owned(),
+        )),
+        Some(i) => Ok(i),
+        None => Ok(-1),
+    }?;
+    Ok(u32::try_from(last_mined_index.saturating_add(GAP_LIMIT)).unwrap())
+}
+
+/// Utility function to return an `InclusiveRange<u32>` that starts at `i + 1`
+/// and is of length up to `n`. The range is truncated if necessary to end at
+/// the maximum valid address index, `i32::MAX`.
+///
+/// Precondition: `i >= -1 and n > 0`
+pub(crate) fn range_after(i: i32, n: i32) -> RangeInclusive<u32> {
+    assert!(i >= -1);
+    assert!(n > 0);
+    let first = u32::try_from(i64::from(i) + 1).unwrap();
+    let last = u32::try_from(i.saturating_add(n)).unwrap();
+    first..=last
+}
+
+/// Returns the ephemeral transparent IVK for a given account ID.
+pub(crate) fn get_ephemeral_ivk<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    account_id: AccountId,
+) -> Result<EphemeralIvk, SqliteClientError> {
+    Ok(get_account(conn, params, account_id)?
+        .ok_or(SqliteClientError::AccountUnknown)?
+        .ufvk()
+        .and_then(|ufvk| ufvk.transparent())
+        .ok_or(SqliteClientError::UnknownZip32Derivation)?
+        .derive_ephemeral_ivk()?)
+}
+
+/// Returns a vector with all ephemeral transparent addresses potentially belonging to this wallet.
+/// If `for_detection` is true, this includes addresses for an additional GAP_LIMIT indices.
+pub(crate) fn get_reserved_ephemeral_addresses<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    account_id: AccountId,
+    for_detection: bool,
+) -> Result<HashMap<TransparentAddress, Option<TransparentAddressMetadata>>, SqliteClientError> {
+    let mut stmt = conn.prepare(
+        "SELECT address, address_index FROM ephemeral_addresses WHERE account_id = :account ORDER BY address_index",
+    )?;
+    let mut rows = stmt.query(named_params! { ":account": account_id.0 })?;
+
+    let mut result = HashMap::new();
+    let mut first_unused_index: Option<i32> = Some(0);
+
+    while let Some(row) = rows.next()? {
+        let addr_str: String = row.get(0)?;
+        let raw_index: u32 = row.get(1)?;
+        first_unused_index = i32::try_from(raw_index)
```

Comment:
Oh. So `NonHardenedChildIndex::from_index` takes a signed integer? That's _awful_.

Reply:
No, `NonHardenedChildIndex::from_index` takes a `u32`. Note that it takes `raw_index`, not `first_unused_index`.

That being said, what we need here are 31-bit integers, which we can either get with a `u32` and an `i < (1 << 31)` check (as in `NonHardenedChildIndex::from_index`), or an `i32` and a `!i.is_negative()` check.